pkg/ccl/changefeedccl/changefeed_processors.go

```diff
@@ -1941,16 +1948,214 @@ func (cf *changeFrontier) manageProtectedTimestamps(
 		}
 	}()
 
+	var ptsEntries changefeedpb.ProtectedTimestampRecords
+	if err := readChangefeedJobInfo(ctx, perTableProtectedTimestampsFilename, &ptsEntries, txn, cf.spec.JobID); err != nil {
+		return false, err
+	}
+	pts := cf.FlowCtx.Cfg.ProtectedTimestampProvider.WithTxn(txn)
+
+	if usePerTableTracking && usePerTablePTS {
+		newPtsTimestamp, updatedPerTablePts, err := cf.managePerTableProtectedTimestamps(ctx, txn, progress, &ptsEntries)
+		if err != nil {
+			return false, err
+		}
+		return cf.advanceProtectedTimestampToTimestamp(ctx, txn, progress, pts, newPtsTimestamp, updatedPerTablePts)
+	}
+
+	if len(ptsEntries.ProtectedTimestampRecords) > 0 {
+		// If we're not using per table protected timestamps to protect lagging tables,
+		// we will rely on the single protected timestamp record to protect all tables,
+		// but we need to release all the per table PTS records that we created earlier.
+		if err := cf.releasePerTableProtectedTimestampRecords(ctx, txn, &ptsEntries); err != nil {
+			return false, err
+		}
+	}
+
+	highwater := cf.frontier.Frontier()
+	if highwater.Less(cf.highWaterAtStart) {
+		highwater = cf.highWaterAtStart
+	}
+	return cf.advanceProtectedTimestampToTimestamp(ctx, txn, progress, pts, highwater, false /* updatedPerTablePts */)
+}
+
+func (cf *changeFrontier) managePerTableProtectedTimestamps(
+	ctx context.Context,
+	txn isql.Txn,
+	progress *jobspb.ChangefeedProgress,
+	ptsEntries *changefeedpb.ProtectedTimestampRecords,
+) (newPtsTimestamp hlc.Timestamp, updatedPerTablePts bool, err error) {
+	// TODO: This seems weird, come back to it. Is there a better way to do this?
```

Comment:
We discussed this offline and one alternative is just to set the "bucket" PTS to be `current time - bucketing interval` so that we never have to worry about regressing it